src/daft-logical-plan/src/optimization/rules/reorder_joins/join_graph.rs

```diff
@@ -0,0 +1,734 @@
+    /// (e.g. an aggregation, an outer join, a source node etc).
+    /// It keeps track of the following state:
+    /// - join_conds_to_resolve: Join conditions (left_on/right_on) from downstream joins that need to be resolved by
+    ///                          linking to some upstream relation.
+    /// - final_name_map: Map from a column's current name in the query plan to its name in the final output schema.
+    ///
+    /// Joins that added conditions to `join_conds_to_resolve` will pop them off the stack after they have been resolved.
+    /// Combining each of their resolved `left_on` conditions with their respective resolved `right_on` conditions produces
+    /// a join edge between the relation used in the left condition and the relation used in the right condition.
+    fn process_node<'a>(
+        &mut self,
+        plan: &'a LogicalPlanRef,
+        join_conds_to_resolve: &mut Vec<(String, LogicalPlanRef, bool)>,
+        final_name_map: &mut HashMap<String, String>,
+    ) {
+        let schema = plan.schema();
+        let mut schema_names: HashSet<_> = schema.names().into_iter().collect();
+        for (name, node, done) in join_conds_to_resolve.iter_mut() {
+            if !*done && schema_names.contains(name) {
+                *node = plan.clone();
+            }
+        }
+        match &**plan {
+            LogicalPlan::Project(Project {
+                input, projection, ..
+            }) => {
+                // Get the mapping from input->output for projections that don't need computation.
+                let projection_input_mapping = projection
+                    .iter()
+                    .filter_map(|e| e.input_mapping().map(|s| (e.name(), s)))
+                    .collect::<HashMap<&'a str, _>>();
+                // To be able to reorder through the current projection, all unresolved columns must either have a
+                // zero-computation projection, or must not be projected by the current Project node (i.e. it should be
+                // resolved from some other branch in the query tree).
+                let reorderable_project = join_conds_to_resolve.iter().all(|(name, _, done)| {
+                    *done
+                        || !schema_names.contains(name)
+                        || projection_input_mapping.contains_key(name.as_str())
+                });
+                if reorderable_project {
+                    for (name, _, done) in join_conds_to_resolve.iter_mut() {
+                        if !*done {
+                            if let Some(new_name) = projection_input_mapping.get(name.as_str()) {
+                                // Remove the current name from the list of schema names so that we can produce
+                                // a set of non-join-key names for the current Project's schema.
+                                schema_names.remove(name);
```

Comment:
we can also achieve this with a book mask

Reply:
Hmm in that case we'd still need a mapping from name to index. I removed the incorrect uses of `schema_names`, but opted to continue keeping track of `non_join_names`